environments/common/src/main/java/org/jboss/weld/environment/deployment/WeldDeployment.java

```diff
@@ -64,4 +87,41 @@ public BeanDeploymentArchive getBeanDeploymentArchive(Class<?> beanClass) {
         }
         return null;
     }
-}
+
+    protected BeanDeploymentArchive createAdditionalBeanDeploymentArchiveIfNeeded(Class<?> beanClass) {
+        WeldBeanDeploymentArchive additionalBda = additionalBeanDeploymentArchives.get(beanClass.getClassLoader());
+        if (additionalBda != null) {
+            additionalBda.addBeanClass(beanClass.getName());
+        } else {
+            additionalBda = createAdditionalBeanDeploymentArchive(beanClass);
+        }
+        return additionalBda;
+    }
+
+    /**
+     * Additional bean deployment archives are used for extentions, synthetic annotated types and beans which do not come from a bean archive.
+     *
+     * @param beanClass
+     * @return the additional bean deployment archive
+     */
+    protected WeldBeanDeploymentArchive createAdditionalBeanDeploymentArchive(Class<?> beanClass) {
+        // At this time only the initial bean class is known
+        Set<String> beanClasses = new HashSet<String>();
+        beanClasses.add(beanClass.getName());
+
+        WeldBeanDeploymentArchive additionalBda = new WeldBeanDeploymentArchive(beanClass.getClassLoader().getClass().getName() + "@"
+                + System.identityHashCode(beanClass.getClassLoader()) + ADDITIONAL_BDA_ID_SUFFIX, beanClasses, null);
```

Comment:
As discussed previously we should probably switch to a simpler single-additional-BDA solution. That would give us the same behavior and does not introduce the bda ID problem.